core/modules/wsched/ChunkTaskList.cc

```diff
@@ -0,0 +1,393 @@
+    }
+    LOGS(_log, LOG_LVL_DEBUG,
+         "ChunkTasks enqueue "
+         << a->getIdStr()
+         << " chunkId=" << _chunkId
+         << " state=" << state
+         << " active.sz=" << _activeTasks._tasks.size()
+         << " pend.sz=" << _pendingTasks.size());
+    if (_activeTasks.empty()) {
+        LOGS(_log, LOG_LVL_DEBUG, "Top of ACTIVE is now: (empty)");
+    } else {
+        LOGS(_log, LOG_LVL_DEBUG, "Top of ACTIVE is now: " << _activeTasks.top()->getIdStr());
+    }
+}
+
+
+/// Set this chunk as the active chunk and move pending jobs to active if needed.
+void ChunkTasks::setActive(bool active) {
+    if (_active != active) {
+        LOGS(_log, LOG_LVL_DEBUG, "ChunkTasks " << _chunkId << " active changed to " << active);
+        if (_active && !active) {
+            movePendingToActive();
+        }
+    }
+    _active = active;
+}
+
+
+/// Move all pending Tasks to the active heap.
+void ChunkTasks::movePendingToActive() {
+    for (auto const& t:_pendingTasks) {
+        LOGS(_log, LOG_LVL_DEBUG, "ChunkTasks " << _chunkId << " pending->active " << t->getIdStr());
+        _activeTasks.push(t);
+    }
+    _pendingTasks.clear();
+}
+
+
+/// @return true if active AND pending are empty.
+bool ChunkTasks::empty() const {
+    return _activeTasks.empty() && _pendingTasks.empty();
+}
+
+
+/// This is ready to advance when _activeTasks is empty and no Tasks are in flight.
+bool ChunkTasks::readyToAdvance() {
+    return _activeTasks.empty() && _inFlightTasks.empty();
+}
+
+
+/// @Return true if a Task is ready to be run.
+// ChunkTasks does not have its own mutex and depends on its owner for thread safety.
+// If a Task is ready to be run, _readyTask will not be nullptr.
+ChunkTasks::ReadyState ChunkTasks::ready(bool useFlexibleLock) {
+    auto logMemManRes =
+        [this](bool starved, std::string const& msg, std::vector<memman::TableInfo> const& tblVect) {
+            setResourceStarved(starved);
+            if (starved) {
+                std::string str;
+                for (auto const& tblInfo:tblVect) {
+                    str += tblInfo.tableName + " ";
+                }
+                LOGS(_log, LOG_LVL_DEBUG, "ready memMan " << msg << " - " << str);
+            }
+        };
+    if (_readyTask != nullptr) {
+        return ChunkTasks::ReadyState::READY;
+    }
+    if (_activeTasks.empty()) {
+        return ChunkTasks::ReadyState::NOT_READY;
+    }
+    // Calling this function doesn't get expensive until it gets here. Luckily,
+    // after this point it will return READY or NO_RESOURCES, and ChunkTaskList::_ready
+    // will not examine any further chunks upon seeing those results.
+    auto task = _activeTasks.top();
+    if (!task->hasMemHandle()) {
+        memman::TableInfo::LockType lckOptTbl = memman::TableInfo::LockType::MUSTLOCK;
+        memman::TableInfo::LockType lckOptIdx = memman::TableInfo::LockType::NOLOCK;
+        if (useFlexibleLock) lckOptTbl = memman::TableInfo::LockType::FLEXIBLE;
```

Comment:
Could you make this an if-else or even ?: instead of setting once and then resetting it later. Sort of odd.